piIOComm.c

```diff
@@ -319,81 +319,22 @@ INT32S piIoComm_sendRS485Tel(INT16U i16uCmd_p, INT8U i8uAddress_p,
 	return 0;
 }
 
-INT32S piIoComm_sendTelegram(SIOGeneric * pRequest_p, SIOGeneric * pResponse_p)
+int piIoComm_sendTelegram(SIOGeneric *msg)
 {
-	INT32S i32uRv_l = 0;
-	INT8U len_l;
+	UIoProtocolHeader *hdr = &msg->uHeader;
 	int ret;
-#if 0				//def DEBUG_DEVICE_IO
-	static INT8U last_out[40][2];
-	static INT8U last_in[40][2];
-#endif
-
-	len_l = pRequest_p->uHeader.sHeaderTyp1.bitLength;
 
-	pRequest_p->ai8uData[len_l] = piIoComm_Crc8((INT8U *) pRequest_p, IOPROTOCOL_HEADER_LENGTH + len_l);
+	ret = pibridge_req_io(hdr->sHeaderTyp1.bitAddress,
+			      hdr->sHeaderTyp1.bitCommand,
+			      msg->ai8uData,
+			      hdr->sHeaderTyp1.bitLength,
+			      NULL,
+			      0);
```

Comment:
Because you're passing a `rcv_len` of 0, <a href="https://github.com/RevolutionPi/linux/blob/devel/revpi-5.10-serdev/drivers/tty/serdev/pibridge.c#L473">pibridge_req_io()</a> emits a warning with `dev_warn_ratelimited()` and always returns `-EBADMSG`.

Thus you'll always emit another warning below and user space will be under the impression that sending the telegram failed even though it succeeded.

Have you actually, uhm... *tested* this change?

You'd need a separate helper in the pibridge library which receives every byte of the response, adds it to the CRC, then throws the byte away. Then after all bytes have been received, validate the CRC.

Or change the semantics of `pibridge_req_io()`, but I'm not sure if that'll have adverse effects in the other places where it's called...

Reply:
I tested this with calibration telegrams and everything worked fine. The response to those telegrams contains zero data bytes, so there is no error when passing a 0 length for the receive buffer. However counter reset telegram responses actually contain data bytes and those lead to the errors you described. I (obviously wrongly) assumed that no telegram response contains any data.
The fix I prefer for this is to provide always a buffer that is large enough. For this the semantic of the pibridge_req_io() api has to be changed in a way that it does not return an error if the passed size of the receive buffer is greater than the size of the received response (which is a strange behaviour anyway).